packages/sdk-node/src/index.ts

```diff
@@ -0,0 +1,123 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+import type { IncomingMessage, ServerResponse } from "http";
+import {
+  GrowthBook,
+  Context,
+  setPolyfills,
+  LocalStorageCompat,
+} from "@growthbook/growthbook";
+
+export {
+  GrowthBook,
+  setPolyfills,
+  configureCache,
+  clearCache,
+} from "@growthbook/growthbook";
+
+export type {
+  Context,
+  Experiment,
+  Result,
+  ExperimentOverride,
+  Attributes,
+  ConditionInterface,
+  ExperimentStatus,
+  FeatureDefinition,
+  FeatureResult,
+  FeatureResultSource,
+  FeatureRule,
+  JSONValue,
+  SubscriptionFunction,
+  Filter,
+  VariationMeta,
+  VariationRange,
+} from "@growthbook/growthbook";
+
+let polyfillsSet = false;
+
+export interface GrowthBookMiddlewareProps<Request extends IncomingMessage> {
+  context: Context;
+  getAttributes?: (
+    req: Request
+  ) => Record<string, unknown> | Promise<Record<string, unknown>>;
```

Comment:
Why are we requiring a `Request` to be passed here? I'm not sure which common Node.js framework this would work with, but in Express, when passing data through the request lifecycle, the recommended pattern is to use the [response locals](http://expressjs.com/en/5x/api.html#res.locals), not the request object (though I know in our backend code we are not following the recommended pattern and have instead augmented an Express request). 

Shouldn't they be able to pass anything in here? Or nothing? They just need a function that returns the attributes, we don't need to enforce the arguments it takes.

We can do something like this instead:

```ts
interface GetAttributes {
    (...args: unknown[]): Promise<Record<string, unknown>>
}
```

This will make it more flexible for the developer's implementation. Of course it means they have to do type assertions but it will not constrain them to a specific API.

### Example without arguments

```ts
const getAttrsNoArgs: GetAttributes = async () => {
    return {
        foo: 'bar'
    }
}
```

### Example using Express' res.locals pattern

```ts
// naive express response type
type ExpressResponse = {
    locals: Record<string, unknown>
}

const getAttrsWithRequestAndResponse: GetAttributes = async (req, res) => {
    const method = (req as Request).method

    const locals = (res as ExpressResponse).locals
    const foo = locals.foo

    return {
        method,
        foo,
    }
}
```

### More complex implementation

This one has a `UserRepository`:

```ts
type User = {
    id: string
    email: string
    registeredAt: string
}

class UserRepository {
    async getUser(): Promise<User> {
        return {
            id: 'user-abc123',
            email: 'user@example.com',
            registeredAt: '1689026908',
        }
    }
}

const getAttrsWithUserRepo: GetAttributes = async (userRepo) => {
    return await (userRepo as UserRepository).getUser()
}
```

Naming it `GetAttributesFunction` might be more inline stylistically for JS/TS than `GetAttributes`.

You can see a [TypeScript playground here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgOITAQTGKwBGArpAM7IDeAUMjcgBQB0TcUA5iQFzKEgDWIAewDuIANoBdAJRcAClAEBbYCQgAeAEoQEAqABNVJXKFYAabn0EiAfFcoBfSpQD0T5AEkFABwA2EBRHA4MGABEGQhYDAACwFiZBZWQn9wEkptEENkVgxsXBIAOQFMNk40HJw8IlJkAF54kgBPEAR6SVqrCmpaKAxCKDCqWiHkGAEBLgByfBYJrpoHB0cXdy9fZLAgkLDCEmNkAFEAD08ekhIJ5FOGbwEEOG8yTyDIfsowBs8UI5OIM80STyhFS1TpDG53B5cTTaPQGIwgUzmfjCEC2RbpTLZLAVEgAdUiUU0AEdCL8sCBdP9ARkIFx0NijFVfiC4I1mvQekSzKc2jUOoNaBiwMh-NEBLoQXROfVkMTSYZJAxRTFdI4hkLkOD7mQ6lLmayDsdTiQqUCIIqtQ85sgNaMBCDLSQGHa1d1ev1QcMaMrxSZrUM7X6hgtHM5XB4fH4AhtgqFuLsEcgSIoMFE9t5gLwUHBkABVFRQTSA3ZgHQNN4fFD56AggU0YC6LiGPAI61+ODAbxN+Gsa09VjKF4QXTYbst3vo7yssjVwsQYuRMuemispotLGzujSZByRTKNSz-n+t1gPoDY9ehuTHbQAC0cHwCAAjAAmADMEyDXuG7c714LAACECHHAqwQAw2gKJ+F7DP2g7QMOo7IBMT4AGwABwAJwAAwvqhOHodB37IA4wb2I4GpYrkUB4gSs5FuMZQMpUxDMnUq7snQN5zoCvJHkMPSnh6cBCB2wpcQWDEyvR84CCWZaKhuBZbvYQA).

Reply:
Not sure I follow.  The middleware is defined once and called for every request.  So the `getAttributes` function needs context about which request it's getting attributes for (e.g. to look up cookie values).  Since we're calling the `getAttributes` function from within the middleware, we need to know exactly which arguments to pass in, so we can't let the user define their own.

Reply:
My understanding would be that the user would define the `getAttributes` function themselves, and the only requirement is that it return attributes as `Record<string, unknown>`. If previous middleware has populated the response locals with relevant data, for example, they wouldn't be able to get them and would need to adapt their app to work with this middleware. This may require them to make another round trip to the DB which wouldn't be ideal. I'm thinking in the case that perhaps they have auth middleware that populates the response locals (as is generally done with the patterns recommended by Express), then if they need to get any data from the response locals, they cannot do it in TypeScript because they are constrained to working with a Request object only. 

For example, this wouldn't pass type checks in TypeScript:

```ts
    const middleware = growthbookMiddleware({
      context: {},
      getAttributes: (res) => {
        return {
          registeredAt: res.locals.registrationDate
        };
      },
    });
```

It seems like we're unnecessarily constraining them to defining things on the request object by enforcing the request be the only argument. It doesn't seem we actually need it to be a request since the user defines the function themselves.